### PR TITLE
Alteration2

### PR DESCRIPTION
**Pull request by hash5683**

### CODE REVIEW

File: authentification/urls.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 from . import views
```
Comment (mhsiddiqui) [inline authentification/urls.py:3]: Import individual views instead of complete module.

File: authentification/views.py [hunk 2]
```diff
@@ -1 +1,27 @@
...
+def user_login(request):
+    '''login page'''
+    if request.method == 'POST':
+        form = LoginForm(request.POST)
+        if form.is_valid():
+            username = form.cleaned_data['username']
+            password = form.cleaned_data['password']
+            user = authenticate(request, username=username, password=password)
+            if user:
+                login(request, user)
```
Comment (mhsiddiqui) [inline authentification/views.py:17]: How you are handling invalid username and password error?

File: registration/forms.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+'''importing django forms'''
+from django.contrib.auth.forms import UserCreationForm
+from django.contrib.auth.models import User
+
+class SignupForm(UserCreationForm):
```
Comment (mhsiddiqui) [inline registration/forms.py:5]: I will prefer you to create your own signup form and handle user creation yourself to understand how user creation works.

File: registration/views.py [hunk 4]
```diff
@@ -1,19 +1,16 @@
...
+# signup page
+def user_signup(request):
+    '''signup page'''
+    if request.method == 'POST':
+        form = SignupForm(request.POST)
+        if form.is_valid():
+            form.save()
             return redirect('login')
-    return render (request,'signup.html')
+    else:
```
Comment (mhsiddiqui) [inline registration/views.py:13]: How you are handling validation errors while creating user?

File: todo/views.py [hunk 5]
```diff
@@ -17,8 +17,35 @@ def createpage(request):
         todo.save()
         return redirect('home')
     return render(request,'create.html')
+
+def delete(request, todo_id):
+    '''delete a todo'''
+    todo=Todo.objects.get(id=todo_id)
```
Comment (mhsiddiqui) [inline todo/views.py]: User should be able to delete only his todo. If user enter a todo which does not exists or does not belongs to him, he must get a validation error.

File: todo/views.py [hunk 6]
```diff
@@ -17,8 +17,35 @@ def createpage(request):
...
+    todo.delete()
+    return redirect('home')
+
+def edit(request, todo_id):
+    '''edit a todo'''
+    todo=Todo.objects.get(id=todo_id)
```
Comment (mhsiddiqui) [inline todo/views.py]: Use form.

File: todo/views.py [hunk 7]
```diff
@@ -17,8 +17,35 @@ def createpage(request):
...
+        desc=request.POST.get('description')
+        todo.title=title
+        todo.description=desc
+        todo.save()
+        return redirect('home')
+    return render(request,'edit.html', {'todo': todo})
+
+def mark_as_done(request, todo_id):
+    '''mark a todo as done'''
+    todo=Todo.objects.get(id=todo_id)
```
Comment (mhsiddiqui) [inline todo/views.py]: Same validation as mentioned for delete.

File: todo/views.py [hunk 8]
```diff
@@ -17,8 +17,35 @@ def createpage(request):
...
+    todo.is_done=True
+    todo.save()
+    return redirect('home')
           
 def submit(request):
```
Comment (mhsiddiqui) [inline todo/views.py]: What is being submitted here?

File: todo/views.py [hunk 9]
```diff
@@ -17,8 +17,35 @@ def createpage(request):
...
     '''Submit button on create page'''
     return render(request,'home.html')
+
+
```
Comment (mhsiddiqui) [inline todo/views.py:42]: Only one empty line.

File: todoProject/urls.py [hunk 10]
```diff
@@ -27,5 +27,8 @@
     path('' , include('registration.urls')),
     path('submit/', views.submit, name='submit'),
     path('create/', views.createpage, name='create'),
-    path('home/', views.homepage, name='home'),   
-]
\ No newline at end of file
+    path('home/', views.homepage, name='home'),
+    path('delete/<int:todo_id>/', views.delete, name='delete'),
```
Comment (mhsiddiqui) [inline todoProject/urls.py]: Move urls to respective apps.

**Review by mhsiddiqui — CHANGES_REQUESTED** [review]

File: todoProject/urls.py (hunk 10 above)
Comment (hash5683) [reply todoProject/urls.py]: Ok

File: todo/views.py (hunk 9 above)
Comment (hash5683) [reply todo/views.py:42]: Ok

File: todo/views.py (hunk 8 above)
Comment (hash5683) [reply todo/views.py]: Ye function erase krna tha 
Jb pehly khud form bnaya tha tb ye use ho ra tha 
Is ko men erase krdeta

File: todo/views.py (hunk 7 above)
Comment (hash5683) [reply todo/views.py]: Ok

File: todo/views.py (hunk 6 above)
Comment (hash5683) [reply todo/views.py]: Acha g

File: todo/views.py (hunk 5 above)
Comment (hash5683) [reply todo/views.py]: Validation add krta hun men

File: registration/forms.py (hunk 3 above)
Comment (hash5683) [reply registration/forms.py:5]: Okay

**Pull request merged**